internal/search/result/intersect.go

```diff
@@ -0,0 +1,117 @@
+package result
+
+import (
+	"sync"
+)
+
+// NewLiveMerger creates a type that will perform a "live" merge on
+// any number of result sets. numSources is the number of result sets that we
+// are merging. Sources are numbered [0,numSources).
+func NewLiveMerger(numSources int) *liveMerger {
+	return &liveMerger{
+		numSources: numSources,
+		matches:    make(map[Key]mergeVal, 100),
+	}
+}
+
+type liveMerger struct {
+	mu         sync.Mutex
+	numSources int
+	matches    map[Key]mergeVal
+}
+
+type mergeVal struct {
+	match       Match
+	sourceMarks []bool
```

Comment:
Premature optimization and naming suggestion. A slice of booleans can be represented as bitset / bitmap.

```suggestion
	seen *roaring.Bitmap
```

Reply:
Any intuition on whether a roaring bitmap would be faster for the standard case of `numSources == 2`? I avoided going that direction just because it seemed like overkill for what will usually be a 2-element slice.

Reply:
Ended up using a bitset (the one used by roaring) rather than a bitmap because our set size will always be quite small, and probably always fit in a single word. The helper methods on it are nice too